lib/choice_type_creator.rb

```diff
@@ -0,0 +1,224 @@
+  # Class for making sure that every resource/profile that has a choice type (that's not
+  # already covered by Synthea) has every choice type represented in the output.
+  # For the moment, this only does Date/time related choice types, because they're easy to
+  # automatically generate.
+  class ChoiceTypeCreator
+    SWAP_COUNT = 1
+    MUST_SUPPORT_CHOICE_TYPES = {
+      FHIR::DiagnosticReport =>
+      {
+        prefix: 'effective',
+        suffixes: %w[DateTime Period],
+        profiles: [
+          'http://hl7.org/fhir/us/core/StructureDefinition/us-core-diagnosticreport-note',
+          'http://hl7.org/fhir/us/core/StructureDefinition/us-core-diagnosticreport-lab'
+        ]
+      },
+      FHIR::Immunization =>
+      {
+        prefix: 'occurrence',
+        suffixes: %w[DateTime String],
+        profiles: [
+          'http://hl7.org/fhir/us/core/StructureDefinition/us-core-immunization'
+        ]
+      },
+      FHIR::Observation =>
+      {
+        prefix: 'effective',
+        suffixes: %w[DateTime Period],
+        profiles: [
+          'http://hl7.org/fhir/StructureDefinition/resprate',
+          'http://hl7.org/fhir/StructureDefinition/heartrate',
+          'http://hl7.org/fhir/StructureDefinition/headcircum',
+          'http://hl7.org/fhir/StructureDefinition/bodyweight',
+          'http://hl7.org/fhir/StructureDefinition/bp',
+          'http://hl7.org/fhir/us/core/StructureDefinition/us-core-smokingstatus',
+          'http://hl7.org/fhir/us/core/StructureDefinition/pediatric-weight-for-height',
+          'http://hl7.org/fhir/us/core/StructureDefinition/us-core-observation-lab',
+          'http://hl7.org/fhir/us/core/StructureDefinition/pediatric-bmi-for-age',
+          'http://hl7.org/fhir/us/core/StructureDefinition/us-core-pulse-oximetry',
+          'http://hl7.org/fhir/StructureDefinition/bodyheight',
+          'http://hl7.org/fhir/StructureDefinition/bodytemp'
+        ]
+      },
+      FHIR::Procedure =>
+      {
+        prefix: 'performed',
+        suffixes: %w[DateTime Period],
+        profiles: [
+          'http://hl7.org/fhir/us/core/StructureDefinition/us-core-procedure'
+        ]
+      }
+    }.freeze
+
+    # Goes through the set of bundles, and ensure we have at least SWAP_COUNT resources
+    # in each resource/profile pair listed above that conform to each choice type.
+    # @param bundles [Array] an array of FHIR::Bundle objects
+    # @return nil
+    def self.check_choice_types(bundles)
+      # Get the resources sorted by their resource type (fhir_models classname)
+      resource_map = get_resources_by_class(bundles)
+      resource_map.each do |resource_klass, resources|
+        # Figure out if this resource is one that we need to add choice types for
+        choice = MUST_SUPPORT_CHOICE_TYPES[resource_klass]
+        next unless choice
+
+        # Loop through each profile that this choice type applies to
+        choice[:profiles].each do |profile_url|
```

Comment:
I think it would be nice if everything in here was done by an instance rather than the class. So `ChoiceTypeCreator.check_choice_types` iterates over the profiles and instantiates an instance of `ChoiceTypeCreator` to do the real work. An instance provides a lot more options for cleaning up the code (e.g. breaking up this big method by replacing comments explaining what code does with methods whose names say what the code does).